### PR TITLE
[OYPD-451] Update daxko, & opney_socrates, add openy_prgf_pgm_reg & openy_programs_search with configuration.

### PR DESCRIPTION
**Pull request by Kerby70**
Jira https://propeople-us.atlassian.net/browse/OYPD-451
D.o https://www.drupal.org/node/2884784

Make sure these boxes are checked before asking for review of your pull request - thank you!

## General checks
- [x] All coding styles are fulfilled and there are no any issues reported by CodeSniffer CI. <br/>CI code sniffer errors
- [x] All tests are running and there are no failed tests reported by CI. <br/>Behat test results
- [x] [Documentation](https://github.com/ymcatwincities/openy/tree/8.x-1.x/docs) has been updated according to PR changes.
- [x] [Steps for review](https://github.com/ymcatwincities/openy/pull/94#issue-204580200) have been provided according to PR changes. <br/>Steps for review
- [x] Make sure you've provided all necessary hook\_update\_N to [support upgrade path](https://github.com/ymcatwincities/openy/blob/8.x-1.x/docs/Development/Upgrade%20path.md).
- [x] Make sure your git email is associated with account on drupal.org, otherwise you won't get commits there. <br/>drupal.org email
- [x] If you would like to get credits on drupal.org, [check documentation](https://github.com/ymcatwincities/openy/blob/8.x-1.x/docs/Development/Contributing.md#drupalorg-credits).

Thank you for your contribution!

## Steps for review
- [x] As an admin configure connection to Daxko at `/admin/config/services/daxko`
- [x] Configure program registration configuration at `/admin/config/services/daxko/programs-search`
- [x] Add Program Registration paragraph to landing page
- [x] See Expected form on page.
- [x] Stepping through form results in the appropriate options & final link

### CODE REVIEW

**Comment by gundevel** [comment]
Build comment file: 
--------------
===============
 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build1389 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build1389 |
| Installation process | http://install.openy-dev.ffwua.com/build1389/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER/1389/

**Comment by gundevel** [comment]
Behat test results file URL: http://ci.openymca.org/build1389/profiles/contrib/openy/build/reports/behat

**Comment by Kerby70** [comment]
retest this please

**Comment by gundevel** [comment]
Build comment file: 
--------------
===============
 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build1396 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build1396 |
| Installation process | http://install.openy-dev.ffwua.com/build1396/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER/1396/

**Comment by gundevel** [comment]
Behat test results file URL: http://ci.openymca.org/build1396/profiles/contrib/openy/build/reports/behat

File: modules/custom/openy_programs_search/src/DataStorage.php [hunk 1]
```diff
@@ -0,0 +1,905 @@
...
+        if (isset($locsr[$loc])) {
+          $programMap[$program->id][$locsr[$loc]] = $locsr[$loc];
+        }
+        else {
+          $ranking = [];
+          foreach ($locexp as $k => $exps) {
+
+            $ranking[$k] = 0;
+            if (substr($program->name, 0, 1) == substr($exps[0], 0, 1)) {
+              if ((strpos($program->name, "South Lake Houston") !== FALSE)) {
```
Comment (alexschedrov) [inline modules/custom/openy_programs_search/src/DataStorage.php]: Potential place to remove. It's hardcoded fr Houston.

File: modules/custom/openy_programs_search/src/DataStorage.php [hunk 2]
```diff
@@ -0,0 +1,905 @@
...
+    $cid = __METHOD__;
+    if ($cache = $this->cache->get($cid)) {
+      $data = $cache->data;
+    }
+    else {
+      $locations_data = $branches = $this->client->getBranches(['limit' => 100]);;
+      $data = [];
+      // @todo Make this configurable.
+      // @todo Should we really skip that?
+      $skip_branches = [
```
Comment (alexschedrov) [inline modules/custom/openy_programs_search/src/DataStorage.php]: Probably should be moved to configuration.

File: docs/Content structure/Paragraphs/Program Registration.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# Program Registration
```
Comment (alexschedrov) [inline docs/Content structure/Paragraphs/Program Registration.md]: Let's call in `Program Registration (Daxko)`

**Review by alexschedrov — CHANGES_REQUESTED** [review]

File: modules/custom/openy_programs_search/src/DataStorage.php (hunk 1 above)
Comment (Kerby70) [reply modules/custom/openy_programs_search/src/DataStorage.php]: Better code comments would help :( All I know is 'South Lake Houston' was being given -10 less weight. I'm guessing the functionality could apply differently for each site, maybe needing multiple programs to have different weight adjustments. But this just subtracts 10 it doesn't pin the item to the top, it's a bit confusing.

File: modules/custom/openy_programs_search/config/install/openy_programs_search.settings.yml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+default_locations: []
+client_id: ''
+domain: ''
+base_url: ''
+registration_path: ''
+get_schools_by_program_path: ''
+get_categories_path: ''
+get_map_categories_by_branch_path: ''
```
Comment (Kerby70) [inline modules/custom/openy_programs_search/config/install/openy_programs_search.settings.yml]: @Sanchiz Should I go ahead and set these to the assumed defaults given in the form descriptions?

File: modules/custom/openy_programs_search/src/DataStorage.php (hunk 1 above)
Comment (alexschedrov) [reply modules/custom/openy_programs_search/src/DataStorage.php]: Let's keep as is and then, once we have time, we can make this as enhancement.

File: modules/custom/openy_programs_search/config/install/openy_programs_search.settings.yml (hunk 4 above)
Comment (alexschedrov) [reply modules/custom/openy_programs_search/config/install/openy_programs_search.settings.yml]: @Kerby70 Please clarify this. I'm not sure that I quite get that.

**Comment by gundevel** [comment]
Build comment file: 
--------------
===============
 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build1520 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build1520 |
| Installation process | http://install.openy-dev.ffwua.com/build1520/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER/1520/

**Comment by gundevel** [comment]
Behat test results file URL: http://ci.openymca.org/build1520/profiles/contrib/openy/build/reports/behat

**Comment by Kerby70** [comment]
After looking at the work on the GroupExPro Schedules Block I now understand why the field was visible on the form display and this should be restored.

**Comment by gundevel** [comment]
Build comment file: 
--------------
===============
 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build1588 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build1588 |
| Installation process | http://install.openy-dev.ffwua.com/build1588/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER/1588/

**Comment by gundevel** [comment]
Behat test results file URL: http://ci.openymca.org/build1588/profiles/contrib/openy/build/reports/behat

**Comment by gundevel** [comment]
Build comment file: 
--------------
===============
 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build1589 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build1589 |
| Installation process | http://install.openy-dev.ffwua.com/build1589/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER/1589/

**Comment by gundevel** [comment]
Behat test results file URL: http://ci.openymca.org/build1589/profiles/contrib/openy/build/reports/behat

**Comment by Kerby70** [comment]
retest this please

**Comment by gundevel** [comment]
Build comment file: 
--------------
===============
 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build1591 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build1591 |
| Installation process | http://install.openy-dev.ffwua.com/build1591/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER/1591/

**Comment by gundevel** [comment]
Behat test results file URL: http://ci.openymca.org/build1591/profiles/contrib/openy/build/reports/behat

**Comment by Kerby70** [comment]
retest this please

**Comment by gundevel** [comment]
Build comment file: 
--------------
===============
 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build1614 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build1614 |
| Installation process | http://install.openy-dev.ffwua.com/build1614/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER/1614/

**Comment by gundevel** [comment]
Behat test results file URL: http://ci.openymca.org/build1614/profiles/contrib/openy/build/reports/behat

**Comment by Sanchiz** [comment]
retest this please

**Comment by gundevel** [comment]
Build comment file: 
--------------
===============
 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build1618 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build1618 |
| Installation process | http://install.openy-dev.ffwua.com/build1618/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER/1618/

**Comment by gundevel** [comment]
Behat test results file URL: http://ci.openymca.org/build1618/profiles/contrib/openy/build/reports/behat

**Comment by alexschedrov** [comment]
1. I'm not able to add paragraph to the page https://screencast.com/t/oj9lz4xjl8
2. Steps for review miss step where we should configure connection to Daxko.

**Comment by gundevel** [comment]
Build comment file: 
--------------
===============
 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build1647 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build1647 |
| Installation process | http://install.openy-dev.ffwua.com/build1647/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER/1647/

**Comment by gundevel** [comment]
Behat test results file URL: http://ci.openymca.org/build1647/profiles/contrib/openy/build/reports/behat

**Comment by Sanchiz** [comment]
retest this please

**Comment by gundevel** [comment]
Build comment file: 
--------------
===============
 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build1652 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build1652 |
| Installation process | http://install.openy-dev.ffwua.com/build1652/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER/1652/

**Comment by gundevel** [comment]
Behat test results file URL: http://ci.openymca.org/build1652/profiles/contrib/openy/build/reports/behat

**Comment by alexschedrov** [comment]
1. Update password field from textfield to password field
image

2. Please convert to upper case
image

3. I've specified credentials for Daxko(see comment in JIRA ticket), but paragraph still doesn't work.
Could you please provide credentials in JIRA ticket that work?

**Comment by gundevel** [comment]
Build comment file: 
--------------
===============
 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build1664 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build1664 |
| Installation process | http://install.openy-dev.ffwua.com/build1664/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER/1664/

**Comment by gundevel** [comment]
Decoupling report http://ci.openymca.org/build1664

**Comment by gundevel** [comment]
Behat test results file URL: http://ci.openymca.org/build1664/profiles/contrib/openy/build/reports/behat

**Comment by Sanchiz** [comment]
retest this please

**Comment by gundevel** [comment]
Build comment file: 
--------------
===============
 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build1678 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build1678 |
| Installation process | http://install.openy-dev.ffwua.com/build1678/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER/1678/

**Comment by gundevel** [comment]
Decoupling report http://ci.openymca.org/build1678

**Comment by gundevel** [comment]
Behat test results file URL: http://ci.openymca.org/build1678/profiles/contrib/openy/build/reports/behat

**Comment by Sanchiz** [comment]
retest this please

**Comment by gundevel** [comment]
Build comment file: 
--------------
===============
 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build1683 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build1683 |
| Installation process | http://install.openy-dev.ffwua.com/build1683/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER/1683/

**Comment by gundevel** [comment]
Decoupling report http://ci.openymca.org/build1683

**Comment by gundevel** [comment]
Behat test results file URL: http://ci.openymca.org/build1683/profiles/contrib/openy/build/reports/behat

**Comment by Sanchiz** [comment]
retest this please

**Comment by bjwilkins** [comment]
Build comment file: 
--------------
===============
 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build1690 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build1690 |
| Installation process | http://install.openy-dev.ffwua.com/build1690/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER/1690/

**Comment by bjwilkins** [comment]
Decoupling report http://ci.openymca.org/build1690

**Comment by bjwilkins** [comment]
Behat test results file URL: http://ci.openymca.org/build1690/profiles/contrib/openy/build/reports/behat

**Review by alexschedrov — APPROVED** [review]

**Pull request merged**